src/service/CountryAPI.ts

```
@@ -1,25 +1,75 @@
 import { BASE_API_URL } from '@/constants';
-import { CountrySchema, type ICountry } from '@/interfaces';
+import {
+  CountrySchema,
+  DetailedCountriesSchema,
+  type IGetCountriesResponse,
+} from '@/interfaces';
 import z from 'zod';
 
-export async function getCountry(params: string = 'all'): Promise<ICountry[]> {
-  const url = new URL(params, BASE_API_URL);
+export async function getCountries(
+  params: string = 'all'
+): Promise<IGetCountriesResponse> {
+  const url = new URL(`v3.1/${params}`, BASE_API_URL);
+
   if (params === 'all') {
-    url.searchParams.set('fields', 'name, flags, capital, area, borders');
+    url.searchParams.set('fields', 'name,flags,capital,area,borders');
   }
   try {
     const response: Response = await fetch(url);
+
     if (!response.ok) throw new Error(response.statusText);
     const data: unknown = await response.json();
-
     const result = z.array(CountrySchema).safeParse(data);
+    if (response.status === 404) {
+      return {
+        countries: [],
+      };
+    }
+    if (!result.success) {
+      return {
+        error: `Invalid data format ${result.error.message}`,
+        countries: [],
+      };
+    }
+    return {
+      countries: result.data,
+    };
+  } catch (error: unknown) {
+    if (error instanceof Error)
+      return {
+        error: error.message,
+        countries: [],
+      };
+    return {
+      error: `unknown error`,
+      countries: [],
+    };
+  }
+}
+export async function getCountry(countryName: string = '') {
+  const url = new URL(`v3.1/name/${countryName}`, BASE_API_URL);
+  try {
+    const response: Response = await fetch(url);
+
+    if (!response.ok) throw new Error(response.statusText);
+    const data: unknown = await response.json();
+
+    const result = z.array(DetailedCountriesSchema).safeParse(data);
+
+    if (response.status === 404) {
+      return null;
+    }
```

If the `response.ok` is false, then that means the response status is not within the 200-299 range.

If you get a `404` response it will throw instead of `return null;`

You should move this before you check if `!response.ok` if you expect this behaviour

---

I think i rewrite it many times and loose the point =)